### DISCUSSION TITLE
Unable to receive Topic structure in Azure IoT Hub

### DISCUSSION THREAD

**Discussion by ksaye** [arendst/Tasmota#11772] [Q&A] [answered]
### PROBLEM DESCRIPTION
Unable to receive Topic structure in Azure IoT Hub.  Azure IoT Hub only supports the topic ``devices/{device_id}/messages/events/``, as discussed here: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#sending-device-to-cloud-messages.

I am only able to see the following message in the IoT Hub:
![image](https://user-images.githubusercontent.com/15837044/115128233-fa21c400-9fa1-11eb-8adf-5a5d48e9964e.png)

``
{
  "body": "Online",
  "enqueuedTime": "2021-04-17T22:07:34.962Z",
  "properties": {
    "mqtt-retain": "true",
    "LWT": ""
  }
}
``

It appears to me that messages such as ``/cmnd/POWER`` and all others are not received, I guess because of the hierarchy in the MQTT structure.
![image](https://user-images.githubusercontent.com/15837044/115127950-1f153780-9fa0-11eb-83d1-771842778890.png)

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [X] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [X] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [X] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [X] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [X] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [X] Device used (e.g., Sonoff Basic): _Huzzah Feather (ESP8266)____
- [X] Tasmota binary firmware version number used: _Project tasmota Tasmota Version 9.3.1.3(tasmota)-2_7_4_9(2021-04-17T16:58:34)____
  - [ ] Pre-compiled
  - [#] Self-compiled
- [X] Flashing tools used: _Tasmotizer____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

Following the steps here: https://github.com/arendst/Tasmota/issues/3686#issuecomment-821892451, I was able to enable TLS and configure it to connect.

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

I would hope to see all the messages sent, but I think the topic hierarchy is causing an issue.  Is there a way to either flatten the hierarchy or use properties or the body of the JSON message?

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

I tried to modify the MqttPublishLib(const char* topic, bool retained) as shown below, but my C is not the best and of course it just reset.

```
bool MqttPublishLib(const char* topic, bool retained) {
  retained = true;
  // If Prefix1 equals Prefix2 disable next MQTT subscription to prevent loop
  if (!strcmp(SettingsText(SET_MQTTPREFIX1), SettingsText(SET_MQTTPREFIX2))) {
    char *str = strstr(topic, SettingsText(SET_MQTTPREFIX1));
    if (str == topic) {
      TasmotaGlobal.mqtt_cmnd_blocked_reset = 4;  // Allow up to four seconds before resetting residual cmnd blocks
      TasmotaGlobal.mqtt_cmnd_blocked++;
    }
  }

/*
  char* topicCopy = strdup(topic);
  #if defined(USE_MQTT_AZURE_IOT)
  // Azure currently does not suport deep topics like 'devices/%deviceid%/messages/events/cmnd/POWER'
  // converting this to 'devices/%deviceid%/messages/events/cmnd-POWER'
  int slashcounter=0;
  while(strtok(topicCopy, "/")) {
      slashcounter++;
  }
  if (slashcounter > 4){
    int baseTopicLen = strlen(SettingsText(SET_MQTT_FULLTOPIC));
    int fullTopicLen = strlen(topic);
    int position = 1;
    while (position <= fullTopicLen){
      if (position > baseTopicLen){
        if (topicCopy[position] == '/'){
          topicCopy[position] = '-';  
        } else {
          topicCopy[position] = topicCopy[position];  
        }
      } else {
        topicCopy[position] = topicCopy[position];
      }
      position++;
    }
  }
  #endif
  const char *publishedTopic = topicCopy;
  free(topicCopy);
*/
  //bool result = MqttClient.publish(publishedTopic, TasmotaGlobal.mqtt_data, retained);
  bool result = MqttClient.publish(topic, TasmotaGlobal.mqtt_data, retained);
  yield();  // #3313
  return result;
}
```

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by ksaye** [comment]
So making the following change in xdrv_02_mqtt.ino gives me the data, but I don't have the original topic.  Any suggestion how I can append the ```const char* topic``` into ```TasmotaGlobal.mqtt_data```?

```
bool MqttPublishLib(const char* topic, bool retained) {
  // If Prefix1 equals Prefix2 disable next MQTT subscription to prevent loop
  if (!strcmp(SettingsText(SET_MQTTPREFIX1), SettingsText(SET_MQTTPREFIX2))) {
    char *str = strstr(topic, SettingsText(SET_MQTTPREFIX1));
    if (str == topic) {
      TasmotaGlobal.mqtt_cmnd_blocked_reset = 4;  // Allow up to four seconds before resetting residual cmnd blocks
      TasmotaGlobal.mqtt_cmnd_blocked++;
    }
  }

  #if defined(USE_MQTT_AZURE_IOT)
    const char* iothub_publish_endpoint = SettingsText(SET_MQTT_FULLTOPIC);
    bool result = MqttClient.publish(iothub_publish_endpoint, TasmotaGlobal.mqtt_data, retained);
  #else
    bool result = MqttClient.publish(topic, TasmotaGlobal.mqtt_data, retained);
  #endif

  yield();  // #3313
  return result;
}
```
And I see it in Azure IoT Hub:
![image](https://user-images.githubusercontent.com/15837044/115130416-488c8e00-9fb5-11eb-8701-8fd48f01906d.png)

**Comment by barbudor** [comment, marked as answer]
Tasmota is designed to work with generic MQTT brokers like Mosquitto.

I played a bit with Adafruit IO which also enforce a specific topic structure. The best solution being to use rules to publish on custom topics. But that's not very convenient.

If you insist in running a cloud broker, you can have a look at Flespi or MyQttHub which provide full featured MQTT brokers with free accounts for limited number of devices (around 50 If I remember).
I've tested successfully both of them.

I know that Tasmota supports also AWS but I have no idea how it works.

**Comment by ksaye** [comment]
All, so after reviewing the Tasmota code, the approach was pretty simple. Because Azure IoT Hub does not accept custom topics (yet), I just set the Tasota topic as a property in the message before sending. When sending a message from IoT Hub to Tasmota, I sent the topic (example cmnd/tasmota/power) as a property and have modified the Tasmota code to read the property into the topic.

I have working code and it is all enabled by defining USE_MQTT_TLS and (new)USE_MQTT_AZURE_IOT in user_config_override.h. I will submit a PR soon -- once I figure out the best branch and approach.

**Reply by pwandrag** [reply]
Hi, thanks for this. I am able to successfully flow telemetry through the hub on Azure IoT with my devices. Have you also looked into the "device twin" construct in Azure IoT? It seems it is a bit more involved than just topic structures.

**Comment by ksaye** [comment]
Glad you like it and it works.

Twins are pretty simple, as it is just a topic to subscribe and publish to.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Reply by pwandrag** [reply]
Hi, thanks for replying, yes I figured as much under the covers, but surely there needs to be some board logic to take the state from the twin and apply it accross all the parameters? the idea with twins is to set the cloud state and when a device wakes up, syncs the state and applies any differences locally

**Reply by ksaye** [reply]
@pwandrag , I agree.  let me know if you want to collaborate on that.

**Reply by pwandrag** [reply]
Hi, sure. Sounds like something that might give some 'enterprise' capabilites to the stack. My C skills are super rusty though, haven't cut a line in 25 years so not sure how much help I can be. This sdk is at least in C and has some bones to work with I guess, if we port some fo the logic. https://github.com/Azure/azure-sdk-for-c/tree/main/sdk/samples/iot/aziot_esp8266 I used to work for Microsoft, so I might be able to get some contacts to help get tot he IoT team there if we need more info that the docs don't give
